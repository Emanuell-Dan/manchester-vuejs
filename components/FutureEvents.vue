<template>
  <div 
    id="future-events"
    class="future-events relative pt-16"
  >
    <h2 class="future-events__heading font-bold text-center text-2xl md:text-4xl mb-4 md:mb-8">
      Coming next
    </h2>
    <div 
      v-if="event"
      class="future-events__card"
    >
      <a 
        :href="event.link"
        target="_blank"
        rel="noreferrer"
        :aria-label="`Upcoming event: ${event.name}`"
        :title="`Upcoming event: ${event.name}`"
        class="future-events__url bg-white flex flex-col md:flex-row rounded-lg max-w-3xl m-auto p-8"
      >
        <div class="future-events__event-info flex-grow">
          <h3 class="future-events__event-name font-bold pb-4">{{ event.name }}</h3>
          <p class="future-events__event-attendees">{{ event.yes_rsvp_count }} {{ getAttendeesLimit }} attendees</p> 
        </div>
        <div class="future-events__event-logistics">
          <p class="future-events__event-date font-medium pb-2">{{ getDateAndTime(event.time) }}</p>
          <p class="future-events__event-venue">{{ event.venue.name }}</p>
          <p class="future-events__event-address">{{ event.venue.address_1 }}</p>
          <p class="future-events__event-address">{{ event.venue.address_2 }}</p>
        </div>
      </a>
    </div>
    <h3 
      v-else
      class="future-events__tba text-center font-bold"
    >
      Watch this space...
    </h3>
  </div>
</template>

<script>
import dateAndTime from '../mixins/dateAndTime.js';

export default {
  name: 'FutureEvents',
  mixins: [dateAndTime],
  props: {
    event: {
      type: Object,
      default: () => {}
    }
  },
  computed: {
    getAttendeesLimit() {
      return this.event && this.event.rsvp_limit ? `/ ${this.event.rsvp_limit}` : '';
    }
  }
};
</script>
