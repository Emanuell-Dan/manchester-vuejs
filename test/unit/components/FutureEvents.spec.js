import FutureEvents from '@/components/FutureEvents.vue';
import { mount } from 'vue-test-utils';

describe('FutureEvents.vue', () => {
	let cmp;

	beforeEach(() => {
		cmp = mount(FutureEvents, {
			propsData: {
				event: {
					name: 'vue.js meetup',
					rsvp_limit: 10,
					yes_rsvp_count: 2,
					time: 1530811800000,
					venue: {
						name: 'Spaces',
						address_1: '57 Spring Gardens',
						address_2: 'Manchester'
					}
				}
			}
		});
	});

	it('should render with expected html structure', () => {
		expect(cmp.element).toMatchSnapshot();
	});

	describe('No event', () => {
		beforeEach(() => {
			cmp.setProps({ event: null});
		});
		it('does not render future event card', () => {
			expect(cmp.find('.future-events__card').exists()).toBeFalsy();
		});

		it('displays correct text', () => {
			expect(cmp.find('.future-events__tba').text()).toBe('Watch this space...');
		});
	});

	describe('has event', () => {
		it('renders future event card', () => {
			expect(cmp.find('.future-events__card').exists()).toBeTruthy();
		});

		it('displays event name', () => {
			expect(cmp.find('.future-events__event-name').exists()).toBeTruthy();
			expect(cmp.find('.future-events__event-name').text()).toBe('vue.js meetup');
		});

		describe('number of attendees', () => {
			it('displays number of attendees with limit', () => {
				expect(cmp.find('.future-events__event-attendees').text()).toBe('2 / 10 attendees');
			});

			it('does not display limit', () => {
				cmp.setProps({ event: { ...cmp.vm.event, rsvp_limit: null }});
				expect(cmp.find('.future-events__event-attendees').text()).toBe('2  attendees');
			});
		});

		describe('displays correct event logistics', () => {
			it('renders venue name', () => {
				expect(cmp.find('.future-events__event-venue').text()).toBe('Spaces');
			});

			it('renders event time', () => {
				expect(cmp.find('.future-events__event-date').text()).toBe('Jul 05 2018 at 18:30');
			});

			it('renders address line 1', () => {
				expect(cmp.find('.future-events__event-address').text()).toBe('57 Spring Gardens');
			});

			it('renders address line 2', () => {
				expect(cmp.find('.future-events__event-address:last-child').text()).toBe('Manchester');
			});
		});
	});
});
