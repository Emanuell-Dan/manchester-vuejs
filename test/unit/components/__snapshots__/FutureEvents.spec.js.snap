// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FutureEvents.vue should render with expected html structure 1`] = `
<div
  class="future-events relative pt-16"
  id="future-events"
>
  <h2
    class="future-events__heading font-bold text-center text-2xl md:text-4xl mb-4 md:mb-8"
  >
    
    Coming next
  
  </h2>
   
  <h3
    class="future-events__tba text-center font-bold"
  >
    
    Watch this space...
  
  </h3>
</div>
`;
